libyul/YulName.cpp

```diff
@@ -0,0 +1,435 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/YulName.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <libyul/optimiser/NameCollector.h>
+
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/AST.h>
```

Comment:
We usually group the includes in alphabetical order. 

```suggestion
#include <libyul/YulName.h>

#include <libyul/backends/evm/EVMDialect.h>
#include <libyul/optimiser/NameCollector.h>
#include <libyul/AST.h>
#include <libyul/Dialect.h>
#include <libyul/Exceptions.h>
```

Reply:
Whatever y'all think makes sense. The [coding style](https://github.com/ethereum/solidity/blob/develop/CODING_STYLE.md#12-include-headers) doesn't state it that way and even doesn't respect it in its own example (`utility` before `numeric`). :shrug: 
If you ask me, ordering alphabetically may even break the order by generality. Like a "common"-style header is likely more general than some implementation header in the same namespace / include hierarchy but is pretty far up in the alphabet. :D
In the end I guess most important is that the headers are self-contained. The rest should more or less follow.

Reply:
Yeah, I agree with you. I honestly don't know if this is really a coding style rule of the team. I just adopted it based on what I saw in the code, but you are right, and maybe we should define a proper rule for that based on generality or something like that.

Reply:
purely based on generality may sometimes be difficult to determine, especially when dealing with 3rd party libraries :) honestly I don't think it matters all that much as long as the first include of a compilation unit is of the header that it implements to ensure no critical information is missing from the header. so ordering alphabetically in the same include hierarchy and otherwise by generality is reasonable to me. Albeit i find the alphabetical ordering a bit pedantic but that's just personal taste :)

Reply:
I don't think this was ever an official rule. Some people ordered their includes that way, some didn't. It actually started to be used more consistently only in recent times, when there was a big reshuffle in the team. This rule seems popular in coding styles and looks like the people who joined liked it more than former team members :)

Personally, I don't mind that much either way, in the past I tried to stick mostly to the order of generality as this makes the most sense to me. But, like you said, that order is not always clear so when someone prefers the alphabetical order I don't object. I actually do like that it's unambiguous and I started ordering my includes that way. I may have even started to point it out non-alphabetical order in some places under that influence.

But in the end I agree with @clonker. It's not very important and there isn't much gain in enforcing it strictly. Though, in the end, if we go with a formatter, it's likely that we'll end up with an alphabetical order anyway :)